client/allochealth/tracker.go

```diff
@@ -420,8 +420,19 @@ OUTER:
 		for _, treg := range allocReg.Tasks {
 			for _, sreg := range treg.Services {
 				for _, check := range sreg.Checks {
-					if check.Status == api.HealthPassing {
+					onupdate := sreg.CheckOnUpdate[check.CheckID]
+					switch check.Status {
+					case api.HealthPassing:
 						continue
+					case api.HealthWarning:
+						if onupdate == structs.OnUpdateIgnoreWarn || onupdate == structs.OnUpdateIgnore {
```

Comment:
This `watchConsulEvents` is long-lived. Will this break ongoing health tracking after a deployment is complete? I'd be nice to have an integration test that walks over this lifecycle.

Reply:
+1 to an integration test. `watchConsulEvents` I believe needs a more accurate comment. It is "long-lived" only in the context of determining health for an allocation, once it has been healthy for the `MinHealhtyTime` it exists. So if an allocation is marked healthy, and then later the health check starts failing, Nomad won't do anything about it, which I thought was interesting https://github.com/hashicorp/nomad/blob/master/client/allochealth/tracker.go#L391

Reply:
Added an e2e test, I can look into an integration level test with a mocked out consul client but am unsure if there's any precedent for that level of testing yet

Reply:
The e2e test is great, but it's not comprehensive with the interactions for ongoing health tracking. If you take a look at [`command/agent/consul/check_watcher_test.go`](https://github.com/hashicorp/nomad/blob/master/command/agent/consul/check_watcher_test.go) you may find a good basis for Consul client dependent tests.